dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/netty/jfapchannel/JMSClientInboundHandler.java

```diff
@@ -0,0 +1,133 @@
+    }
+	
+	
+	protected final static AttributeKey<OutboundConnection> CONNECTION_KEY = AttributeKey.valueOf("OutboundConnection");
+	protected final static AttributeKey<String> CHAIN_ATTR_KEY = AttributeKey.valueOf("CHAIN_NAME");
+	
+	/** Called when a new connection is established */
+	@Override
+	public void channelActive(ChannelHandlerContext ctx) throws Exception {
+		if (tc.isEntryEnabled())
+            SibTr.entry(this, tc, "channelActive", ctx.channel());
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			SibTr.debug(this, tc, "channelActive", ctx.channel().remoteAddress() + " connected for chain " + ctx.channel().attr(CHAIN_ATTR_KEY).get());
+		}
+		// Set attribute to point to the appropriate connection
+		
+		Attribute<OutboundConnection> attr = ctx.channel().attr(CONNECTION_KEY);
+		if(attr.get() == null) {
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				SibTr.debug(this, tc, "channelActive", "could not associate an incoming conection with a Connection.");
+            }
+		}
+		if (tc.isEntryEnabled())
+            SibTr.exit(this, tc, "channelActive", ctx.channel());
+
+	}
+
+	@Override
+	protected void channelRead0(ChannelHandlerContext ctx, WsByteBuffer msg) throws Exception {
+		if (tc.isEntryEnabled())
+            SibTr.entry(this, tc, "channelRead0", ctx.channel());
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			SibTr.debug(this, tc, "channelRead0", ctx.channel() + ". [" + msg.array() + "] bytes received");
+		}
+		
+		
+		Attribute<OutboundConnection> attr = ctx.channel().attr(CONNECTION_KEY);
+		OutboundConnection connection = attr.get();
+
+		//TODO: Check if connection is closed
+        if (connection != null) {
+        	IOReadCompletedCallback callback = connection.getReadCompletedCallback();
+        	IOReadRequestContext readCtx = connection.getReadRequestContext();
+        	NetworkConnection networkConnection = connection.getNetworkConnection();
+        	if(
+        			callback instanceof NettyConnectionReadCompletedCallback && 
+        			readCtx instanceof NettyIOReadRequestContext && 
+        			networkConnection instanceof NettyNetworkConnection) {
+        		((NettyConnectionReadCompletedCallback)callback).readCompleted(msg, readCtx, (NettyNetworkConnection)networkConnection);
+        	}else {
+        		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+        			SibTr.debug(this, tc, "channelRead0: Something's wrong. Callback or read context is not netty specific. Cry cause not sure what happened.", new Object[] {connection, callback, readCtx});
+                }
+        	}
+        	
+        } else {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            	SibTr.debug(this, tc, "channelRead0", "could not associate an incoming message with a Connection. Message will be ignored.");
```

Comment:
Should we be considering these paths as an error condition that should lead the connection to close (exchange to finish)? Seems like something is corrupt at this point